### DISCUSSION TITLE
Is there a way to synthesize invariants of a specific form？

### DISCUSSION THREAD

**Discussion by whuzhaoyx** [cvc5/cvc5#10253] [Q&A] [answered]
I am a new user of this artistic tool to synthesize invariants. In Z3, we can synthesize the invariant with specific form:  inv= a1*(n+a2)**2 + a3 - x<=0, where a1, a2, a3 is the real values i want to resolve, and x1, x2 x3 are vars. How can i specify the form of invariants in CVC5? Could you please help me with that and provide some code? Many thanks!

Z3 code:
``` python
n,x,a1,a2,a3 = Reals('n x a1 a2 a3')
xrange = And(x>=-10, x<=10, n>=-10,n<=10)
arange = And(a1>=-2, a1<=2, a2>=-2, a2<=2, a3>=-2, a3<=2)

inv = a1*(n+a2)**2 + a3 - x<=0


guard1 = -1<=0
inv_f1 = a1*(n+1+a2)**2 + a3 - (x+6*n+6)<=0

loop_cond = n-5<=0
pre = And(x==0,n==0)
post = And(2.5*n**2+4*n-x<=0,x-3.5*n**2-3*n<=0)

cons1 = ForAll([x,n],Implies(And(pre, xrange), inv))
cons2 = ForAll([x,n],Implies(And(inv, guard1, loop_cond, xrange), inv_f1))
cons3 = ForAll([x,n],Implies(And(inv, Not(loop_cond),xrange),post))
               
s = Solver()
s.add( cons1, cons2, cons3, arange)
print(s.check())
print(s.model())
```
In CVC5 , i don't know how to do that. The following code can't specify the form of invariants i want.

``` python
 post_f = slv.defineFun("post-f", [x1, x2], boolean, post)
 inv_f = slv.synthFun("inv-f", [x1, x2], boolean)
 # declare the invariant-to-synthesize

 slv.addSygusInvConstraint(inv_f, pre_f, trans_f, post_f)
```

**Comment by ajreynol** [comment, marked as answer]
I believe you should use the "grammar" field of `synthFun` method. An example of using grammars in the API is in this example: https://cvc5.github.io/docs/cvc5-1.1.0/examples/sygus-fun.html, see lines 62-66 for how to define the grammar, which is passed to `synthFun` on line 70.

**Reply by whuzhaoyx** [reply]
Ok! Thank u very much.
